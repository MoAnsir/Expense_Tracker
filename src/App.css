.App {
  text-align: center;
}

.App h1 {
  background-color: #282c34;
  color: white;
  font-size: 7rem
}



