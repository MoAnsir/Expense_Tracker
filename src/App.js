import { Container } from "react-bootstrap";
import AddExpense from "./Comps/AddExpense";
import BreakdownExpense from "./Comps/BreakdownExpense";
import FilterExpense from "./Comps/FilterExpense";
import ListExpense from "./Comps/ListExpense";
import Styles from "./App.module.scss";
import { useState } from "react";
import { v4 as uuidv4 } from "uuid";

function App({ testInitialState }) {
    const initialState = [
        {
            id: uuidv4(),
            date: new Date(),
            desc: "initial",
            amount: 0,
            type: "initial",
        },
    ];

    const [expenses, setExpenses] = useState(
        testInitialState ? testInitialState : initialState
    );

    //const [expenses, setExpenses] = useState(initialState);

    return (
        <div className={Styles.app}>
            <h1 className="p-4">Expense tracker</h1>
            <Container>
                <AddExpense setExpenses={setExpenses} expenses={expenses} />
                <FilterExpense expenses={expenses} setExpenses={setExpenses} />
                <ListExpense setExpenses={setExpenses} expenses={expenses} />
                <BreakdownExpense expenses={expenses} />
            </Container>
        </div>
    );
}

export default App;
