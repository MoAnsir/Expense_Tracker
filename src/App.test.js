import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";

import App from "./App";

afterEach(cleanup);

describe("Expense Tracker App", () => {
    it("should render the components", () => {
        render(<App />);

        const mainHeading = screen.getByRole("heading", {
            name: /Expense tracker/i,
        });
        const addExpenseComp = screen.getByTestId("add-expense");
        const filterComp = screen.getByTestId("filter-expense");
        const listExpense = screen.getByTestId("list-expense");
        const expenseChart = screen.getByTestId("breakdown-expense");

        expect(mainHeading).toBeInTheDocument();
        expect(addExpenseComp).toBeInTheDocument();
        expect(filterComp).toBeInTheDocument();
        expect(listExpense).toBeInTheDocument();
        expect(expenseChart).toBeInTheDocument();
    });

    it("should add a new expense", async () => {
        render(<App />);

        //const { getByPlaceholderText, getByRole } = render(<App />);

        const expenseDescription =
            screen.getByPlaceholderText(/Expense description/i);
        const expenseAmount = screen.getByPlaceholderText(/Amount/i);
        const expenseType = screen.getByRole("combobox", {
            name: /Type of expense:/i,
        });
        const addExpenseButton = screen.getByRole("button", {
            name: /Add Expense/i,
        });

        expect(expenseDescription).toBeInTheDocument();
        expect(expenseAmount).toBeInTheDocument();
        expect(screen.getAllByRole("option").length).toBe(9);
        expect(expenseType).toBeInTheDocument();
        expect(addExpenseButton).toBeInTheDocument();

        fireEvent.change(expenseDescription, {
            target: { value: "test expense" },
        });
        fireEvent.change(expenseAmount, { target: { value: "1000" } });
        fireEvent.change(expenseType, {
            target: { value: "mortgage" },
        });

        expect(expenseDescription.value).toBe("test expense");
        expect(expenseAmount.value).toBe("1000");
        expect(expenseType.value).toBe("mortgage");

        fireEvent.click(addExpenseButton);
        expect(expenseDescription.value).toBeTruthy();
        expect(expenseAmount.value).toBeFalsy();
        expect(screen.queryByText(/test expense/i)).not.toBeInTheDocument();
    });

    it("should edit an expense", () => {
        render(<App />);
    });

    it("should delete an expense", () => {
        render(<App />);
    });

    it("should filter the expenses", () => {
        render(<App />);
    });
});

describe("Jest Snapshot testing suite", () => {
    it("Matches DOM Snapshot", () => {
        const domTree = renderer.create(<App />).toJSON();
        expect(domTree).toMatchSnapshot();
    });
});
