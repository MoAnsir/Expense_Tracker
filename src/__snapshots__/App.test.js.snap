// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Snapshot testing suite Matches DOM Snapshot 1`] = `
<div
  className="app"
>
  <h1
    className="p-4"
  >
    Expense tracker
  </h1>
  <div
    className="container"
  >
    <div
      className="mb-5 row"
      data-testid="add-expense"
    >
      <h2
        className="p-2"
      >
        Add Expense
      </h2>
      <input
        className="p-2 mb-2"
        onChange={[Function]}
        placeholder="Expense description"
        type="text"
        value=""
      />
      <input
        className="p-2 mb-2"
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value=""
      />
      <label
        className="mb-2"
        htmlFor="expense"
      >
        Type of expense:
      </label>
      <select
        className="p-2"
        id="expense"
        name="expense"
        onChange={[Function]}
        value=""
      >
        <option
          value="select"
        >
          Please Select
        </option>
        <option
          value="mortgage"
        >
          Mortgage
        </option>
        <option
          value="insurance"
        >
          Insurance
        </option>
        <option
          value="food"
        >
          Food
        </option>
        <option
          value="clothing"
        >
          Clothing
        </option>
        <option
          value="utilities"
        >
          Utilities
        </option>
        <option
          value="transport"
        >
          Transport
        </option>
        <option
          value="entertainment"
        >
          Entertainment
        </option>
        <option
          value="other"
        >
          Other
        </option>
      </select>
      <button
        className="my-2 btn btn-primary"
        disabled={false}
        name="add_expense"
        onClick={[Function]}
        type="button"
      >
        Add Expense
      </button>
    </div>
    <div
      className="mb-5 row"
      data-testid="filter-expense"
    >
      <div
        className="col"
      >
        <div>
          <h2>
            Filter expenses
          </h2>
          <button
            className="btn btn-primary"
            disabled={false}
            name="time"
            onClick={[Function]}
            type="button"
          >
            Time
          </button>
          <button
            className="mx-5 btn btn-primary"
            disabled={false}
            name="amount"
            onClick={[Function]}
            type="button"
          >
            Amount
          </button>
          <button
            className="btn btn-primary"
            disabled={false}
            name="type"
            onClick={[Function]}
            type="button"
          >
            Type
          </button>
        </div>
      </div>
    </div>
    <div
      className="mb-5 text-center row"
      data-testid="list-expense"
    >
      <div
        className="p-0 m-0 col"
      >
        <h2>
          Current Expenses
        </h2>
        <ul
          className="p-0 m-0"
        >
          <li
            className=""
          >
            <button
              className="me-4 btn btn-primary"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Delete
            </button>
            <p
              className="d-inline"
            >
              initial
            </p>
            <p
              className="d-inline mx-4"
            >
              Â£
              0
            </p>
            <p
              className="d-inline"
            >
              initial
            </p>
            <button
              className="ms-4 btn btn-primary"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Edit
            </button>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="row"
      data-testid="breakdown-expense"
    >
      <div
        className="col"
      >
        <canvas
          height={150}
          role="img"
          width={300}
        />
      </div>
    </div>
  </div>
</div>
`;
